hs_core/views/utils.py

```diff
@@ -559,84 +559,125 @@ def get_metadata_contenttypes():
-                                  then=Value(True, BooleanField()))))
+def get_my_resources_list(user, annotate=False, filter=None, **kwargs):
+    """
+    Gets a QuerySet object for listing resources that belong to a given user.
+    :param user: an instance of User - user who wants to see his/her resources
+    :param annotate: whether to annotate for my resources page listing.
+    :return: an instance of QuerySet of resources
+    """
+    resource_collection = BaseResource.objects.none()
+    owned_resources = BaseResource.objects.none()
+    editable_resources = BaseResource.objects.none()
+    viewable_resources = BaseResource.objects.none()
+    discovered_resources = BaseResource.objects.none()
+
+    # if annotate:  # When used in the My Resources page, annotate for speed
+    if not filter or 'owned' in filter:
+        # get a list of resources with effective OWNER privilege
+        owned_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.OWNER)
+        # remove obsoleted resources from the owned_resources
+        owned_resources = owned_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id')).exclude(extra_data__to_be_deleted__isnull=False)
+
+    if not filter or 'editable' in filter:
+        # get a list of resources with effective CHANGE privilege (should include resources that the
+        # user has access to via group
+        editable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.CHANGE,
+                                                                             via_group=True)
+        # remove obsoleted resources from the editable_resources
+        editable_resources = editable_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id'))
+
+    if not filter or 'viewable' in filter:
+        # get a list of resources with effective VIEW privilege (should include resources that the
+        # user has access via group
+        viewable_resources = user.uaccess.get_resources_with_explicit_access(PrivilegeCodes.VIEW,
+                                                                             via_group=True)
+        # remove obsoleted resources from the viewable_resources
+        viewable_resources = viewable_resources.exclude(object_id__in=Relation.objects.filter(
+            type='isReplacedBy').values('object_id'))
+
+    if not filter or 'discovered' in filter:
+        discovered_resources = user.ulabels.my_resources
+        if annotate:
+            discovered_resources = discovered_resources.all().annotate(discovered=Value(True, BooleanField()))
+
+    favorite_resources = user.ulabels.favorited_resources
+
+    # join all queryset objects. if/else saves one database query
+    if not filter or 'is_favorite' in filter:
+        resource_collection = owned_resources.distinct() | \
```

Comment:
I am wondering when the `is_favorite` is on if we can do this:
`resource_collection = resource_collection |  favorite_resources.distinct()`